arangod/Transaction/Helpers.cpp

```diff
@@ -433,6 +443,469 @@ std::string transaction::helpers::makeIdFromParts(
+    if (!options.keepNull && s.isObject()) {
+      b->add(VPackValue(it.first));
+      VPackCollection::merge(*b, VPackSlice::emptyObjectSlice(), s, true, true);
+    } else {
+      b->addUnchecked(it.first, s);
+    }
+
+    if (batchOptions.computedValues != nullptr) {
+      keysWritten.emplace(it.first);
+    }
+  }
+
+  b->close();
+
+  if (batchOptions.computedValues != nullptr) {
+    // add all remaining computed attributes, if we need to
+    batchOptions.ensureComputedValuesContext(trx, collection);
+    batchOptions.computedValues->mergeComputedAttributes(
+        *batchOptions.computedValuesContext, trx, b->slice(), keysWritten,
+        ComputeValuesOn::kUpdate, builder);
+  } else {
+    // add document as is
+    builder.add(b->slice());
+  }
+
+  TRI_ASSERT(revisionId.isSet());
+  return {};
+}
+
+/// @brief new object for insert, computes the hash of the key
+Result transaction::helpers::newObjectForInsert(
+    transaction::Methods& trx, LogicalCollection& collection,
+    velocypack::Slice value, RevisionId& revisionId,
+    velocypack::Builder& builder, OperationOptions const& options,
+    transaction::BatchOptions& batchOptions) {
+  transaction::BuilderLeaser b(&trx);
+
+  b->openObject();
+
+  // add system attributes first, in this order:
+  // _key, _id, _from, _to, _rev
+
+  // _key
+  VPackSlice s = value.get(StaticStrings::KeyString);
+  if (s.isNone()) {
+    TRI_ASSERT(!options.isRestore);  // need key in case of restore
+    auto keyString = collection.keyGenerator().generate(value);
+
+    if (keyString.empty()) {
+      return Result(TRI_ERROR_ARANGO_OUT_OF_KEYS);
+    }
+
+    b->add(StaticStrings::KeyString, VPackValue(keyString));
+  } else if (!s.isString()) {
+    return Result(TRI_ERROR_ARANGO_DOCUMENT_KEY_BAD);
+  } else {
+    TRI_ASSERT(s.isString());
+
+    // validate and track the key just used
+    auto res = collection.keyGenerator().validate(s.stringView(), value,
+                                                  options.isRestore);
+
+    if (res != TRI_ERROR_NO_ERROR) {
+      return Result(res);
+    }
+
+    b->add(StaticStrings::KeyString, s);
+  }
+
+  // _id
+  uint8_t* p = b->add(StaticStrings::IdString,
+                      VPackValuePair(9ULL, VPackValueType::Custom));
+
+  *p++ = 0xf3;  // custom type for _id
+
+  if (trx.state()->isDBServer() && !collection.system()) {
+    // db server in cluster, note: the local collections _statistics,
+    // _statisticsRaw and _statistics15 (which are the only system
+    // collections)
+    // must not be treated as shards but as local collections
+    encoding::storeNumber<uint64_t>(p, collection.planId().id(),
+                                    sizeof(uint64_t));
+  } else {
+    // local server
+    encoding::storeNumber<uint64_t>(p, collection.id().id(), sizeof(uint64_t));
+  }
+
+  // _from and _to
+  if (collection.type() == TRI_COL_TYPE_EDGE) {
+    bool extendedNames = trx.vocbase().server().hasFeature<DatabaseFeature>() &&
```

Comment:
NIT
Is `trx.vocbase().server().hasFeature<DatabaseFeature>()` always true?
But still think get server once will be look better